hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -41,11 +41,11 @@ object PruneDeadFields {
               false
           }
         case (t1: TTuple, t2: TTuple) =>
-          t1.required == t2.required &&
+          (!t1.required || t2.required) &&
             t1.size == t2.size &&
             t1.types.zip(t2.types)
               .forall { case (elt1, elt2) => isSupertype(elt1, elt2) }
-        case (t1: Type, t2: Type) => t1 == t2
+        case (t1: Type, t2: Type) => t1 == t2 || t1.setRequired(true) == t2
```

Comment:
this should be `t1 == t2 || t2.required && t2.isOfType(t1)`

setRequired can be very expensive